### PR TITLE
🦺 zv: *Fd now wrappers around the new safe std API

### LINKED ISSUE #286: Port file descriptor related API to Rust's safe-io
In GitLab by @msandova on Nov 3, 2022, 19:58

It would be interesting to make use of the newer safe-io APIs in Rust. It in not clear to what extent one can do this without breaking API.

### PR DESCRIPTION
**Pull request by zeenix**
This meant having to make many `Value` conversions fallible. I tried very hard to not have to do that but in the end, that's the price we must pay for providing an ergonomic safe API. This doesn't affect high-level code much though since `Value` and `OwnedValue` derives have been updated to implement `TryFrom` instead of `From` in appropriate places.

Fixes #286.

### CODE REVIEW

File: zvariant/src/value.rs [hunk 1]
```diff
@@ -230,18 +231,18 @@ impl<'a> Value<'a> {
...
-                let o = OwnedValue::from(&**v);
+                let o = OwnedValue::try_from(&**v)?;
                 Value::Value(Box::new(o.into_inner()))
             }
 
-            Value::Array(v) => Value::Array(v.to_owned()),
-            Value::Dict(v) => Value::Dict(v.to_owned()),
-            Value::Structure(v) => Value::Structure(v.to_owned()),
+            Value::Array(v) => Value::Array(v.try_to_owned()?),
+            Value::Dict(v) => Value::Dict(v.try_to_owned()?),
```
Comment (A6GibKm) [inline zvariant/src/value.rs:239]: Why were these changed? If `v` implemented to_owned in std, then this should be possible to fail.

File: zvariant/src/fd.rs [hunk 2]
```diff
@@ -1,22 +1,105 @@
...
+    /// Borrow the file descriptor.
+    pub fn borrow(&self) -> BorrowedFd<'_> {
+        match self {
+            Self::Borrowed(fd) => fd.as_fd(),
+            Self::Owned(fd) => fd.as_fd(),
+        }
+    }
+
+    /// Try to create an owned version of `self`.
+    pub fn try_to_owned(&self) -> crate::Result<Fd<'static>> {
```
[the remark this answers is not in the document]
Comment (zeenix) [reply zvariant/src/fd.rs:31]: hmm.. good point.

File: zvariant/src/value.rs [hunk 3]
```diff
@@ -273,6 +274,38 @@ impl<'a> Value<'a> {
...
+            Value::U32(v) => Value::U32(*v),
+            Value::I64(v) => Value::I64(*v),
+            Value::U64(v) => Value::U64(*v),
+            Value::F64(v) => Value::F64(*v),
+            Value::Str(v) => Value::Str(v.clone()),
+            Value::Signature(v) => Value::Signature(v.clone()),
+            Value::ObjectPath(v) => Value::ObjectPath(v.clone()),
+            Value::Value(v) => Value::Value(Box::new(v.try_clone()?)),
+            Value::Array(v) => Value::Array(v.try_clone()?),
+            Value::Dict(v) => Value::Dict(v.try_clone()?),
```
Comment (A6GibKm) [inline zvariant/src/value.rs:300]: Ditto.

File: zvariant/src/fd.rs [hunk 4]
```diff
@@ -1,22 +1,105 @@
...
-pub struct Fd(io::RawFd);
+#[derive(Debug)]
+pub enum Fd<'f> {
+    Borrowed(BorrowedFd<'f>),
+    Owned(fd::OwnedFd),
+}
+
+impl<'f> Fd<'f> {
+    /// Borrow the file descriptor.
+    pub fn borrow(&self) -> BorrowedFd<'_> {
```
[the remark this answers is not in the document]
Comment (zeenix) [reply zvariant/src/fd.rs:23]: hmm.. good point.

File: zvariant/src/fd.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (zeenix) [reply zvariant/src/fd.rs:31]: > Any reason to not have the same api as the rust types (try_clone_to_owned)?

For consistency with our own API. :) Besides `try_clone_to_owned` sounds very strange to me. `ToOwned::to_owned` isn't `CloneToOwned::clone_to_owned`.

File: zvariant/src/fd.rs [hunk 5]
```diff
@@ -1,22 +1,105 @@
...
+impl From<OwnedFd> for Fd<'static> {
+    fn from(owned: OwnedFd) -> Self {
+        owned.inner
+    }
+}
+
+impl From<Fd<'static>> for fd::OwnedFd {
+    fn from(fd: Fd<'static>) -> Self {
+        match fd {
+            Fd::Borrowed(_) => unreachable!("Can't convert a borrowed FD to owned"),
```
[the remark this answers is not in the document]
Comment (zeenix) [reply zvariant/src/fd.rs:68]: > I am not entirely sure if this is truly unreachable,

Hm.. right, one can create a `BorrowedFd<'static>` I guess and then convert that to `Fd::Borrowed` using this impl. Good catch! I'll change to `TryFrom`.

[the remark this answers is not in the document]
Comment (A6GibKm) [reply zvariant/src/fd.rs:68]: e.g. would `Fd::from(unsafe { BorrowedFd::from_raw(27) })` produce a  Fd with static lifetime?

File: zvariant/src/value.rs (hunk 1 above)
Comment (zeenix) [reply zvariant/src/value.rs:239]: Those container types have `Value` contained in them so they had to be converted to be fallible as well.

File: zvariant/src/value.rs (hunk 3 above)
Comment (zeenix) [reply zvariant/src/value.rs:300]: same reason. :)

File: zvariant/src/fd.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (zeenix) [reply zvariant/src/fd.rs:68]: I think so, yes. I recently wrote some code (that I had to ditch) that did that and it ended up with `'static` lifetime even though I didn't intend for it to do that.

[the remark this answers is not in the document]
Comment (zeenix) [reply zvariant/src/fd.rs:68]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7faddd8af1b4c3ae0863c17a74d9b29c

File: zvariant/src/fd.rs [hunk 6]
```diff
@@ -1,33 +1,109 @@
...
+    }
+}
+
+impl From<OwnedFd> for Fd<'static> {
+    fn from(owned: OwnedFd) -> Self {
+        owned.inner
+    }
+}
+
+impl TryFrom<Fd<'static>> for fd::OwnedFd {
```
Comment (A6GibKm) [inline zvariant/src/fd.rs:57]: Now that we clone, could this requirement be soften up to a random lifetime `'fd`?. Besides that, MR looks :ok_hand:  to me.
